src/components/hooks.js

```diff
@@ -231,29 +199,51 @@ export function useClosestVitessceContainerSize(ref) {
   return [width, height];
 }
 
-export function useExpressionValueGetter({ attrs, expressionData }) {
+export function useExpressionValueGetter({ instanceObsIndex, matrixObsIndex, expressionData }) {
   // Get a mapping from cell ID to row index in the gene expression matrix.
-  const cellIdMap = useMemo(() => {
-    const result = {};
-    if (attrs && attrs.rows) {
-      // eslint-disable-next-line no-plusplus
-      for (let i = 0; i < attrs.rows.length; i++) {
-        result[attrs.rows[i]] = i;
-      }
+  // Since the two obsIndices (instanceObsIndex = the obsIndex from obsEmbedding)
+  // may be ordered differently (matrixObsIndex = the obsIndex from obsFeatureMatrix),
+  // we need a way to look up anobsFeatureMatrix obsIndex index
+  // given an obsEmbedding obsIndex index.
+  const toMatrixIndexMap = useMemo(() => {
+    if (instanceObsIndex && matrixObsIndex) {
+      const matrixIndexMap = new Map(matrixObsIndex.map((key, i) => ([key, i])));
+      return instanceObsIndex.map(key => matrixIndexMap.get(key));
     }
-    return result;
-  }, [attrs]);
+    return null;
+  }, [instanceObsIndex, matrixObsIndex]);
 
   // Set up a getter function for gene expression values, to be used
   // by the DeckGL layer to obtain values for instanced attributes.
-  const getExpressionValue = useCallback((entry) => {
-    const cellId = entry[0];
-    if (cellIdMap && expressionData && expressionData[0]) {
-      const cellIndex = cellIdMap[cellId];
-      const val = expressionData[0][cellIndex];
+  const getExpressionValue = useCallback((entry, { index: instanceIndex }) => {
+    if (toMatrixIndexMap && expressionData && expressionData[0]) {
+      const rowIndex = toMatrixIndexMap[instanceIndex];
+      const val = expressionData[0][rowIndex];
       return val;
     }
     return 0;
-  }, [cellIdMap, expressionData]);
+  }, [toMatrixIndexMap, expressionData]);
   return getExpressionValue;
 }
+
+export function useGetObsInfo(obsType, obsIndex, obsLabelsTypes, obsLabelsData) {
+  return useCallback((obsId) => {
+    if (obsId) {
+      return {
+        [`${capitalize(obsType)} ID`]: obsId,
+        ...fromEntries(Object.entries(obsLabelsTypes).map(([scopeKey, obsLabelsType]) => ([
+          obsLabelsType,
+          obsLabelsData?.[scopeKey]?.obsLabels?.[
+            // TODO: Maybe all loaders that return obsIndex should also return an obsIndexMap
+            // with keys: obsId, values: obsIdx
+            // which would avoid the indexOf calls.
```

Comment:
Yes but that would be a bit clunky.....not sure what the alternative would be.  Perhaps (and this would be a big ask) it might make sense to change the index data structures to something like a `set` or `Map`.  They should be unique after all...